### PR TITLE
CG: Initial implementation of Constrained Generics

### PR DESCRIPTION
**Pull request by vasslitvinov**
This PR introduces Constrained Generics (CG) with interfaces and interface constraints.
It provides a straw-person implementation of bare-bones essential functionality.

**Note:** this introduces two new keywords: `interface` and `implements`

Implementation details of a slightly earlier version are presented in the PR #16813,
which is superseded by this PR.

While there:
* Reformatted _BaseAST::astTagAsString()_ for readability.
* Removed a piece of dead code in ResolutionCandidate.cpp

Detailed development history is partly available in:  &nbsp; f48a0a0a31..8d19bb82bd &nbsp; f48a0a0a31..a2021610e1 &nbsp; f48a0a0a31..9764de6a3e

Next steps after the merge:

* [ ] Add the tests from the earlier CG effort, see
    https://github.com/mppf/chapel/tree/chriswailes_ast_integration/test/interfaces

* [ ] The TODOs in compiler comments that are annotated with "CG TODO".

* [ ] Implement interfaces in nested / function scopes.
  For that, we will need to implement, for example, _InterfaceSymbol::copyInner()_

### CODE REVIEW

File: compiler/include/expr.h [hunk 1]
```diff
@@ -275,6 +275,36 @@ class NamedExpr : public Expr {
 };
 
 
+class ImplementsExpr : public Expr {
```
Comment (mppf) [inline compiler/include/expr.h:278]: Can you put a comment above explaining the syntax this class supports?

File: compiler/include/interfaces.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef _INTERFACES_H_
+#define _INTERFACES_H_
+
+// wass do we need this file?
```
Comment (mppf) [inline compiler/include/interfaces.h:24]: apparently not!

File: compiler/include/stmt.h [hunk 3]
```diff
@@ -268,6 +268,38 @@ class GotoStmt : public Stmt {
   LabelSymbol*        gotoTarget()                                     const;
 };
 
+/************************************* | **************************************
+*                                                                             *
+*                                                                             *
+************************************** | *************************************/
+
+class ImplementsStmt : public Stmt {
```
Comment (mppf) [inline compiler/include/stmt.h:276]: let's put a comment explaining the syntax this class supports

File: compiler/include/symbol.h [hunk 4]
```diff
@@ -556,6 +556,33 @@ class TypeSymbol : public Symbol {
 *                                                                             *
 ************************************** | *************************************/
 
+class InterfaceSymbol : public Symbol {
```
Comment (mppf) [inline compiler/include/symbol.h:559]: let's put a comment describing the purpose of this AST class / what syntax it supports

File: compiler/include/symbol.h [hunk 5]
```diff
@@ -755,6 +779,10 @@ extern Symbol *gStringC;
 extern Symbol *gOpaque;
 extern Symbol *gTimer;
 extern Symbol *gTaskID;
+extern Symbol *gDummyWitness;
```
Comment (mppf) [inline compiler/include/symbol.h:782]: Can we have a comment describing what gDummyWitness is for?

File: compiler/include/type.h [hunk 6]
```diff
@@ -364,6 +364,29 @@ class PrimitiveType : public Type {
 };
 
 
+/************************************* | **************************************
+*                                                                             *
+* a ConstrainedType is the type of:                                           *
+* (1) a formal of a CG function if it is subject to interface constraint(s)   *
+* (2) a formal of an InterfaceSymbol if it has the type intent                *
```
Comment (mppf) [inline compiler/include/type.h:371]: Are there any other formals to InterfaceSymbols? Can you have this comment show syntax examples?

File: compiler/include/view.h [hunk 7]
```diff
@@ -36,11 +36,8 @@ Expr*    aidExpr(int id);
 BaseAST* aid(BaseAST* ast);
 Expr*    aidExpr(BaseAST* ast);
 
-void        list_view_noline(const BaseAST* ast);
-void        nprint_view(BaseAST* ast);
-
-// defined in codegen/codegen.cpp
-void        nprint_view(GenRet& gen);
+bool inTree(BaseAST* ast);
+bool inTree(int id);
```
Comment (mppf) [inline compiler/include/view.h:40]: Ideally these changes would be in a different PR as they aren't specific to the CG effort, right?

File: compiler/parser/chapel.ypp [hunk 8]
```diff
@@ -1084,6 +1089,30 @@ $$ = buildIfStmt(convertAssignmentAndWarn($2,$3,$4), $6, $8); }
 $$ = buildIfStmt(convertAssignmentAndWarn($2,$3,$4), $5, $7); }
 ;
 
+interface_stmt:
+  TINTERFACE ident_def TLP ifc_formal_ls TRP block_stmt
+  { $$ = buildChapelStmt(InterfaceSymbol::buildDef($2, $4, $6)); }
```
Comment (mppf) [inline compiler/parser/chapel.ypp:1094]: I don't think there is agreement about putting the parameters always in parens in this way. The CHIP proposal supports a single implicit parameter called Self. Additionally, I don't personally know of a reason we would ever need more than one parameter. I think the language design here needs some discussion, but it doesn't have to hold up this PR. Just don't think that adding this PR means the issue is settled.

File: compiler/parser/chapel.ypp [hunk 9]
```diff
@@ -1084,6 +1089,30 @@ $$ = buildIfStmt(convertAssignmentAndWarn($2,$3,$4), $6, $8); }
...
+| ifc_formal_ls TCOMMA ifc_formal { $1->insertAtTail($3); }
+;
+
+ifc_formal:
+  TTYPE ident_def  { $$ = InterfaceSymbol::buildFormal($2, INTENT_TYPE); }
+;
+
+implements_stmt:
+  TIMPLEMENTS ident_def TLP actual_ls TRP TSEMI
+  { $$ = buildChapelStmt(ImplementsStmt::build($2, $4, NULL)); }
```
Comment (mppf) [inline compiler/parser/chapel.ypp:1108]: The CHIP had support for e.g. `int implements Addable` which you can view as an alternative syntax for `implements Addable(int)`. It seems easy enough to support in this 1st PR - can you add it?

File: compiler/parser/chapel.ypp [hunk 10]
```diff
@@ -1084,6 +1089,30 @@ $$ = buildIfStmt(convertAssignmentAndWarn($2,$3,$4), $6, $8); }
...
+;
+
+implements_expr:
+  TIMPLEMENTS ident_def TLP actual_ls TRP
+  { $$ = ImplementsExpr::build($2, $4); }
```
Comment (mppf) [inline compiler/parser/chapel.ypp:1113]: Same as above - I'd like to see `int implements Addable` as an alternative syntax for `implements Addable(int)`.

File: test/constrained-generics/basic-arg-query.good [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  witness 'reqFun'  basic-arg-query.chpl:33
+  default 'dfltImpl'  basic-arg-query.chpl:22
+
+checking constraint for 'IFC'  basic-arg-query.chpl:27
+  for call to cgFun  at basic-arg-query.chpl:39
+  satisfied with implements stmt  basic-arg-query.chpl:37
+
+considering CG candidate 'cgFun'  basic-arg-query.chpl:27
+  witness(es):
+  implements stmt for ifc 'IFC'  basic-arg-query.chpl:37
```
Comment (mppf) [inline test/constrained-generics/basic-arg-query.good:17]: These don't belong in the the .good file unless you add a developer-only flag to enable the tracing and provide it in the compopts.

File: test/constrained-generics/error-call-ug.good [hunk 12]
```diff
@@ -0,0 +1,10 @@
+resolving interface declaration 'IFC'  error-call-ug.chpl:9
+  required fn 'reqFun'
+  required fn 'dfltImpl'  with a default implementation
+
+resolving CG function early 'cgFun'  error-call-ug.chpl:16
+error-call-ug.chpl:16: In function 'cgFun':
+error-call-ug.chpl:18: error: unresolved call 'ugFun(Q)'
+error-call-ug.chpl:21: note: this candidate did not match: ugFun(arg)
+error-call-ug.chpl:18: note: because call actual argument #1 of an interface type Q
+error-call-ug.chpl:21: note: is passed to a generic formal
```
Comment (mppf) [inline test/constrained-generics/error-call-ug.good:10]: I think this error message should say something about how you can't call an unconstrained generic from a constrained generic.

**Review by mppf — COMMENTED** [review]
I've reviewed the tests, headers, and parser changes so far.

File: compiler/AST/FnSymbol.cpp [hunk 13]
```diff
@@ -780,6 +801,10 @@ bool FnSymbol::hasGenericFormals(SymbolMap* map) const {
     if (formal->intent == INTENT_PARAM) {
       isGeneric = true;
 
+    } else if (isConstrainedType(formal->type)) {
+      if (isConstrainedGeneric())
+        isGeneric = true;
```
Comment (mppf) [inline compiler/AST/FnSymbol.cpp:806]: Should this have an `else` clause to do an `INT_FATAL` ? Is it possible to have a formal type that is a constrained type outside of a constrained generic?

File: compiler/AST/FnSymbol.cpp [hunk 14]
```diff
@@ -118,6 +120,19 @@ void FnSymbol::verify() {
     INT_FATAL(this, "Bad FnSymbol::lifetimeConstraints::parentSymbol");
   }
 
+  // constrainedTypes: AList of DefExpr of ConstrainedType
+  INT_ASSERT(constrainedTypes.parent == this);
+  for_alist(ctExpr, constrainedTypes) {
+    Symbol* ctSym = toDefExpr(ctExpr)->sym;
```
Comment (mppf) [inline compiler/AST/FnSymbol.cpp:126]: maybe assert isDefExpr and then assert isTypeSymbol so we get internal errors rather than core dump if the verify fails?

File: compiler/resolution/visibleFunctions.cpp [hunk 15]
```diff
@@ -225,7 +225,11 @@ void findVisibleFunctions(CallInfo&             info,
   }
 }
 
+/************************************* | *************************************/
+// CG TODO: revert these changes if they end up unused
```
Comment (mppf) [inline compiler/resolution/visibleFunctions.cpp:229]: It looks like they are in fact unused?

File: compiler/AST/primitive.cpp [hunk 16]
```diff
@@ -997,6 +997,7 @@ initPrimitive() {
   prim_def(PRIM_GET_USER_LINE, "_get_user_line", returnInfoDefaultInt, true, true);
   prim_def(PRIM_GET_USER_FILE, "_get_user_file", returnInfoInt32, true, true);
 
+  prim_def(PRIM_PROXY_REF, "proxy reference", returnInfoVoid, false);
```
Comment (mppf) [inline compiler/AST/primitive.cpp:1000]: Can you put a comment here describing the purpose and usage of this new primitive ?

File: compiler/AST/view.cpp [hunk 17]
```diff
@@ -1154,6 +1226,30 @@ void set_view(std::set<BlockStmt*>& bss) {
...
+      countNonNull++;
+      if (!strcmp(ty->name, name)) {
+        count++;
+        printf("  sym %8d   %s %d  %s\n", ty->id, ty->type->astTagAsString(),
+               ty->type->id, debugLoc(ty));
+      }
+    }
+  }
+  printf("  = %d type(s) of %d\n", count, countNonNull);
+}
```
Comment (mppf) [inline compiler/AST/view.cpp:1251]: there's a lot of debug code in here that would be better in another PR

File: compiler/passes/checkParsed.cpp [hunk 18]
```diff
@@ -350,6 +350,7 @@ checkFunction(FnSymbol* fn) {
   // empty body instead.  This is consistent with the current draft
   // of the spec as well.
   if (fn->hasFlag(FLAG_NO_FN_BODY) && !fn->hasFlag(FLAG_EXTERN))
+   if (!isInterfaceSymbol(fn->defPoint->parentSymbol))
```
Comment (mppf) [inline compiler/passes/checkParsed.cpp:353]: Let's use the normal two spaces per indent level, please

File: compiler/passes/normalize.cpp [hunk 19]
```diff
@@ -386,6 +390,39 @@ static bool isInLifetimeClause(CallExpr* call) {
...
+* undergo normalization.
+*
+* TODO handle the case where the 'where' clause has other things
+* besides a single IC, for example 'where IFC1(...) && IFC2(...)'.
+*
+************************************** | *************************************/
+
+static void moveInterfaceConstraints() {
+  forv_Vec(ImplementsExpr, ie, gImplementsExprs) {
+    if (isImplementsStmt(ie->parentExpr))
```
Comment (mppf) [inline compiler/passes/normalize.cpp:408]: Should this check parent's parent etc? In other words, is it possible that an ImplementsStmt contains an ImplementsExpr at some deeper nesting?

File: compiler/AST/primitive.cpp (hunk 16 above)
Comment (mppf) [reply compiler/AST/primitive.cpp:1000]: I see now it has to do with resolution order for constrained generics. Can you rename it to something that indicates that better? E.g. `PRIM_RESOLVE_IN_ORDER` ?

File: compiler/resolution/ResolutionCandidate.cpp [hunk 20]
```diff
@@ -282,6 +310,11 @@ bool ResolutionCandidate::computeSubstitutions(Expr* ctx) {
         computeSubstitutionForDefaultExpr(formal, ctx);
         nDefault++;
       }
+
+      // todo handle param formals too
```
Comment (mppf) [inline compiler/resolution/ResolutionCandidate.cpp:314]: There are a bunch of TODOs around. Could you try making .future tests for most of them?

File: compiler/resolution/ResolutionCandidate.cpp [hunk 21]
```diff
@@ -869,6 +891,15 @@ bool ResolutionCandidate::checkGenericFormals(Expr* ctx) {
         return false;
       }
 
+      // wass does this work for ref formals, ref actuals?
```
Comment (mppf) [inline compiler/resolution/ResolutionCandidate.cpp:894]: seems like a todo here for yourself?

File: compiler/resolution/ResolutionCandidate.cpp [hunk 22]
```diff
@@ -1064,6 +1095,15 @@ void explainCandidateRejection(CallInfo& info, FnSymbol* fn) {
     case RESOLUTION_CANDIDATE_IMPLICIT_WHERE_FAILED:
       USR_PRINT(fn, "because an argument was incompatible");
       break;
+    case RESOLUTION_CANDIDATE_INTERFACE_CONSTRAINTS_NOT_SATISFIED:
+      USR_PRINT(fn, "because interface constraint(s) were not satisfied\n");
```
Comment (mppf) [inline compiler/resolution/ResolutionCandidate.cpp:1099]: Isn't it really that the constrained generic *function*'s constraints were not satisfied? I would expect the error you have here for `int implements Addable` when `int` doesn't have an `add` function, e.g.

In any case I think we would like to show more information about what constraint was not satisfied.

File: compiler/AST/interfaces.cpp [hunk 23]
```diff
@@ -0,0 +1,469 @@
...
+#include "passes.h"
+#include "stmt.h"
+#include "stringutil.h"
+
+/////////////////////////////////////////////////////////////////////////////
+// InterfaceSymbol
+//
+
+//
+// For now an interface formal is a VarSymbol, not a ArgSymbol.
```
Comment (mppf) [inline compiler/AST/interfaces.cpp:33]: a ArgSymbol -> an ArgSymbol

Comment (mppf) [reply compiler/AST/interfaces.cpp:33]: But isn't this comment wrong anyway right now because they are TypeSymbols?

File: compiler/AST/interfaces.cpp [hunk 24]
```diff
@@ -0,0 +1,469 @@
...
+
+  // at least 1 formal is required
+  if (isym->ifcFormals.empty())
+    USR_FATAL_CONT(isym, "interface %s has no formal arguments\n"
+                   "  at least one formal argument is required", isym->name);
+
+  // ifcBody can be empty; check that the content is appropriate
+  for_alist(expr, isym->ifcBody->body)
+    if (toInterfaceFunDecl(expr) == NULL)
+      USR_FATAL_CONT(expr, "not a 'proc' or 'iter' declaration\n"
```
Comment (mppf) [inline compiler/AST/interfaces.cpp:77]: I'd probably a following `USR_PRINT` rather than putting `\n` in the error

File: compiler/AST/interfaces.cpp [hunk 25]
```diff
@@ -0,0 +1,469 @@
...
+void ImplementsStmt::accept(AstVisitor* visitor) {
+  if (visitor->enterImplementsStmt(this)) {
+    implExpr->accept(visitor);
+    implBody->accept(visitor);
+    visitor->exitImplementsStmt(this);
+  }
+}
+
+void ImplementsStmt::replaceChild(Expr* oldAst, Expr* newAst) {
+  // TODO update 'witnesses'
```
Comment (mppf) [inline compiler/AST/interfaces.cpp:271]: Seems like something to do before merging?

File: compiler/AST/interfaces.cpp [hunk 26]
```diff
@@ -0,0 +1,469 @@
...
+  parentFn->interfaceConstraints.insertAtTail(ie);
+
+  // catch up with moveGlobalDeclarationsToModuleScope()
+  if (!wasConstrainedGeneric)
+    if (ModuleSymbol* mod = toModuleSymbol(parentFn->defPoint->parentSymbol))
+      {
+        FnSymbol* initFn = mod->initFn;
+        // follows insertProxyRef()
+        // wass todo this will not retain the exact location,
+        // wass if needed, do desugarInterfaceAsType() earlier in normalize
```
Comment (mppf) [inline compiler/AST/interfaces.cpp:358]: Seems like running it before moveGlobalDeclarationsToModuleScope would make this code simpler.

File: compiler/AST/interfaces.cpp [hunk 27]
```diff
@@ -0,0 +1,469 @@
...
+}
+
+// To handle the cases where 'implExpr' requires resolving,
+// change wrapImplementsStatements as follows.
+// During normalization, set the type of the wrapper function's formal
+// ('arg' below) to dtAny. Detect this case when resolving a call
+// to a wrapper function, see implementsWrapperMatches() //wass correct?
+// There, replace the formal's type with the type of the actual
+// of the (by now hopefully resolved) implements statement.
+static Symbol* computeImplSym(Expr* implExpr) {
```
Comment (mppf) [inline compiler/AST/interfaces.cpp:414]: what's an implSym? Is this trying to compute the symbol for the interface implemented? Could we rename it to better indicate what it is doing?

File: compiler/resolution/interfaceResolution.cpp [hunk 28]
```diff
@@ -0,0 +1,457 @@
...
+
+  populateInterfaceFunctionMap(fn);
+  // Seems like we do not need fn->setGeneric(false) for resolveFunction()
+  // NB we want 'fn' to be generic later, when checking isApplicable()
+  resolveFunction(fn);
+
+  for_alist(expr, fn->constrainedTypes)
+    toDefExpr(expr)->sym->addFlag(FLAG_GENERIC);
+
+  for (int indx = 0; indx < (int)fn->interface2actuals.size(); indx++) {
```
Comment (mppf) [inline compiler/resolution/interfaceResolution.cpp:193]: put a comment like "Remove the functions for the interface that we used during resolution"

does this mean that the calls in the constrained generic will be resolved to functions that aren't in tree anymore? Is that a problem?

**Review by mppf — COMMENTED** [review]
Looking pretty good! Besides the comments I have above you'll have to check testing and for TODOs/wass comments throughout.

**Comment by mppf** [comment]
Can you add the tests from the earlier CG effort? They are here

https://github.com/mppf/chapel/tree/chriswailes_ast_integration/test/interfaces

It would be reasonable to make any that do not work into .futures.

**Review by vasslitvinov — COMMENTED** [review]
@mppf - I got the PR to a close-to-final state.

I responded to your requests in this "review".
Mostly I implemented your suggestions.

File: compiler/parser/chapel.ypp (hunk 8 above)
Comment (mppf) [reply compiler/parser/chapel.ypp:1094]: It is issue #16966.

File: test/constrained-generics/error-call-ug.good (hunk 12 above)
Comment (mppf) [reply test/constrained-generics/error-call-ug.good]: Maybe just adding `note: cannot call an unconstrained generic function from a constrained generic` or something along those lines as an additional line

File: compiler/AST/view.cpp (hunk 17 above)
Comment (mppf) [reply compiler/AST/view.cpp]: I don't think it'd be "trivial" but I'd expect it would be pretty easy to get it reviewed
